### ISSUE TITLE
Support for "message" with "patterns" in `no-restricted-imports`

### ISSUE THREAD

**Issue by daleyjem** [eslint/eslint#11134]
The following issue is exactly what I've run into, but I can't comment an interest in it because comments are restricted to me:
https://github.com/eslint/eslint/issues/9846

Seems like one would want to see a message explaining why a certain pattern is restricted.

**What rule do you want to change?**
no-restricted-imports

**Does this change cause the rule to produce more or fewer warnings?**
Same

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option to create custom messages for pattern-based configurations

**Please provide some example code that this change will affect:**

Consider the following rule:

```js
{
    // No deep imports
    'no-restricted-imports': [1, {patterns: ['*/*', '!@*/*', '!.*/*', '@*/*/*']}]
}
```

It would warn on:

```js
import {thing} from 'some-package/utils'
// or
import {thing} from '@workspace/some-package/utils'
``` 

But pass these:

```js
import {thing} from 'some-package'
// or
import {thing} from '@workspace/some-package'
```

A developer will only see the following from `esw`:
`import is restricted from being used by a pattern no-restricted-imports`

However, I would like to notify them that this import is restricted because it violates an organization-wide "Deep import" rule.

**What does the rule currently do for this code?**
Report an error on each line using the generic message for this rule.

**What will the rule do after it's changed?**
Report an error on each line using the generic message for this rule plus the message supplied in the configuration file.

**Are you willing to submit a pull request to implement this change?**
If I was assured that it'd be merged in, perhaps... but there would be a fair amount of code familiarization.

**Comment by platinumazure** [comment]
Hi @daleyjem, thanks for the issue.

When we originally were adding messaging support to rules like `no-restricted-imports`, we saw that the scenario of supporting messages for patterns required a large option schema change, but weren't sure how commonly it would be used. I asked in the issue if it was something people would use, and at the time it didn't seem like it would be useful for the community, so we decided not to add them at that time.

Now that this issue exists, we can evaluate again and see if it would be useful to others in the community. :smile:

(Note: The fact that the previous issue was closed and comments locked does not mean the original idea is something we will never support. We just like to keep our backlog as focused as we can, and trying to read comments in old issues can be very difficult. The intent in those cases is for people to create a new issue, as you have done, and this will allow the discussion to be much more focused. Thanks for your understanding!)

**Comment by daleyjem** [comment]
@platinumazure Understood. Thank you sir for the prompt attention.

**Comment by daleyjem** [comment]
@nzakas Please see the updated message.

**Comment by osdiab** [comment]
This is closed but it's something that I would love to see.

I have a package with two styles of outputs, `lib/*` is commonJS exports (for a node server that consumes the package), and `es6/*` is ES6 module exports (for tree-shaking on the frontend).

But if you use the `lib/` imports in the web side, you don't get tree-shaking; and if you use the `es6/` imports in the node side, node will crash (but typescript won't tell you you're doing anything wrong).

The `patterns` style rule works perfectly for this, but the error message is less than great - I'd love to tell people to just change the import to the proper subdirectory.

**Comment by g-plane** [comment]
@osdiab Please create a new issue for your request.
